hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/staging/Paths.java

```diff
@@ -0,0 +1,306 @@
+   * @param uuid uuid of job
+   * @return a path which can be used for temporary work
+   * @throws IOException on an IO failure.
+   */
+  public static Path getMultipartUploadCommitsDirectory(Configuration conf,
+      String uuid) throws IOException {
+    FileSystem fs = FileSystem.get(conf);
+    return getMultipartUploadCommitsDirectory(fs, conf, uuid);
+  }
+
+  /**
+   * Build a temporary path for the multipart upload commit information
+   * in the supplied filesystem (which is expected to be the cluster FS)
+   * @param fs target FS
+   * @param conf configuration
+   * @param uuid uuid of job
+   * @return a path which can be used for temporary work
+   * @throws IOException on an IO failure.
+   */
+  static Path getMultipartUploadCommitsDirectory(FileSystem fs,
+      Configuration conf, String uuid) throws IOException {
+    return path(tempDirForStaging(fs, conf),
+        UserGroupInformation.getCurrentUser().getShortUserName(),
+        uuid,
+        STAGING_UPLOADS);
+  }
+
+  // TODO: verify this is correct, it comes from dse-storage
+  private static Path localTemp(Configuration conf, int taskId, int attemptId)
+      throws IOException {
+    String[] dirs = conf.getTrimmedStrings(BUFFER_DIR);
+    Random rand = new Random(Objects.hashCode(taskId, attemptId));
+    String dir = dirs[rand.nextInt(dirs.length)];
+
+    return FileSystem.getLocal(conf).makeQualified(new Path(dir));
+  }
+
+  /**
+   * Returns the partition of a relative file path, or null if the path is a
+   * file name with no relative directory.
+   *
+   * @param relative a relative file path
+   * @return the partition of the relative file path
+   */
+  protected static String getPartition(String relative) {
+    return getParent(relative);
+  }
+
+  /**
+   * Get the set of partitions from the list of files being staged.
+   * This is all immediate parents of those files. If a file is in the root
+   * dir, the partition is declared to be
+   * {@link StagingCommitterConstants#TABLE_ROOT}.
+   * @param attemptPath path for the attempt
+   * @param taskOutput list of output files.
+   * @return list of partitions.
+   * @throws IOException IO failure
+   */
+  public static Set<String> getPartitions(Path attemptPath,
+      List<FileStatus> taskOutput)
+      throws IOException {
+    // get a list of partition directories
+    Set<String> partitions = Sets.newLinkedHashSet();
+    for (FileStatus fileStatus : taskOutput) {
+      // sanity check the output paths
+      Path outputFile = fileStatus.getPath();
+      if (!fileStatus.isFile()) {
+        throw new PathIsDirectoryException(outputFile.toString());
+      }
+      String partition = getPartition(
+          getRelativePath(attemptPath, outputFile));
+      partitions.add(partition != null ? partition : TABLE_ROOT);
+    }
+
+    return partitions;
+  }
+
+  /**
+   * path filter.
+   */
+  static final class HiddenPathFilter implements PathFilter {
```

Comment:
Seems like we write this class over and over. Should we move it into the public API instead of having a local one here?

Reply:
its ubiquitous, isn't it? And in java 8, we can reimplement as a closure...

Reply:
OK, moved up to S3AUtils; combined with other filters in s3a (Listing) and the one in S3ATestUtils.